codalab/bin/ws_server.py

```diff
@@ -1,75 +1,204 @@
+# This is the real ws-server, basically.
 # Main entry point for CodaLab cl-ws-server.
 import argparse
+import json
 import asyncio
+from collections import defaultdict
 import logging
 import re
-from typing import Any, Dict
+from typing import Any, Dict, List
 import websockets
+from dataclasses import dataclass
+import threading
+import time
+
+ACK=b'a'
 
 logger = logging.getLogger(__name__)
-logger.setLevel(logging.WARNING)
+logger.setLevel(logging.INFO)
 logging.basicConfig(format='%(asctime)s %(message)s %(pathname)s %(lineno)d')
 
-worker_to_ws: Dict[str, Any] = {}
 
+"""
+TODO!!! WE NEED TO ADD A SECRET OR SOME SORT OF AUTH FOR THE WS SERVER ENDPOITNS
```

Comment:
Auth, we have two options:
1. merge the rest server and websocket server: either A) switch away from bottle, use flask, or B) maybe just run websocket in a separate process
2. set up some auth for websocket server (maybe it sends a secret initially, or query rest server -> DB to check credentials)